src/io_uring.c

```diff
@@ -0,0 +1,97 @@
+#include "io_uring.h"
+
+#ifdef HAVE_LIBURING
+#include <liburing.h>
+#include <string.h>
+
+#include "serverassert.h"
+#include "zmalloc.h"
+
+#define IO_URING_DEPTH 256 /* io_uring instance queue depth. */
+
+static size_t io_uring_queue_len = 0;
+static io_uring *_io_uring;
+
+void initIOUring(void) {
+    struct io_uring_params params;
+    struct io_uring *ring = zmalloc(sizeof(struct io_uring));
+    memset(&params, 0, sizeof(params));
+    /* On success, io_uring_queue_init_params(3) returns 0 and ring will
+     * point to the shared memory containing the io_uring queues.
+     * On failure -errno is returned. */
+    assert(io_uring_queue_init_params(IO_URING_DEPTH, ring, &params) == 0);
+    _io_uring = ring;
+}
+
+io_uring *getIOUring() {
+    if (_io_uring) {
+        return _io_uring;
+    } else {
+        initIOUring();
+        return _io_uring;
+    }
+}
+
+/* Submit fdatasync request to io_uring. */
+void ioUringPrepFsyncAndSubmit(io_uring *io_uring, int fd) {
+    struct io_uring_sqe *sqe = io_uring_get_sqe(io_uring);
+    io_uring_prep_fsync(sqe, fd, IORING_FSYNC_DATASYNC);
+    io_uring_queue_len++;
+    io_uring_submit(io_uring);
+}
+
+void ioUringWaitFsyncBarrier(io_uring *io_uring) {
+    struct io_uring_cqe *cqe;
+    while (io_uring_queue_len) {
+        if (io_uring_wait_cqe(io_uring, &cqe) == 0) {
+            if (cqe->res < 0) {
+                char errormsg[1024];
+                strerror_r(-cqe->res, errormsg, sizeof(errormsg));
+                fprintf(stderr,
+                        "Can't persist AOF for fsync error when the AOF fsync policy is 'always': %s. Exiting...",
+                        errormsg);
+                exit(1);
```

Comment:
should we do `serverPanic` to get the call stack too?

Reply:
on a second thought, I think we should bubble this error back to the caller and let it decide whether to exit the server or not. In the everysec case, continuing is ok.

Reply:
Update to the print nothing, and return the error code to caller.